web/packages/teleport/src/Roles/RoleEditor/StandardEditor/useStandardModel.ts

```diff
@@ -0,0 +1,197 @@
+};
+type SetResourceAccessAction = {
+  type: 'set-resource-access';
+  payload: ResourceAccess;
+};
+type RemoveResourceAccessAction = {
+  type: 'remove-resource-access';
+  payload: { kind: ResourceAccessKind };
+};
+type AddAccessRuleAction = { type: 'add-access-rule'; payload?: never };
+type SetAccessRuleAction = { type: 'set-access-rule'; payload: RuleModel };
+type RemoveAccessRuleAction = {
+  type: 'remove-access-rule';
+  payload: { id: string };
+};
+type SetOptionsAction = { type: 'set-options'; payload: OptionsModel };
+
+/** Produces a new model using existing state and the action. */
+const reduce = (
+  state: StandardEditorModel,
+  action: StandardModelAction
+): StandardEditorModel => {
+  // We need to give `type` a different name or the assertion in the `default`
+  // block will cause a syntax error.
+  const { type, payload } = action;
+
+  switch (type) {
+    case 'set-role-model':
+      return updateRoleModel(state, payload);
+
+    case 'set-metadata':
+      return patchRoleModel(state, { metadata: payload });
+
+    case 'set-options':
+      return patchRoleModel(state, { options: payload });
+
+    case 'add-resource-access':
+      return patchRoleModel(state, {
+        resources: [
+          ...state.roleModel.resources,
+          newResourceAccess(payload.kind),
+        ],
+      });
+
+    case 'set-resource-access':
+      return patchRoleModel(state, {
+        resources: state.roleModel.resources.map(r =>
+          r.kind === payload.kind ? payload : r
+        ),
+      });
+
+    case 'remove-resource-access':
+      return patchRoleModel(state, {
+        resources: state.roleModel.resources.filter(
+          r => r.kind !== payload.kind
+        ),
+      });
+
+    case 'add-access-rule':
+      return patchRoleModel(state, {
+        rules: [...state.roleModel.rules, newRuleModel()],
+      });
+
+    case 'set-access-rule':
+      return patchRoleModel(state, {
+        rules: state.roleModel.rules.map(r =>
+          r.id === payload.id ? payload : r
+        ),
+      });
+
+    case 'remove-access-rule':
+      return patchRoleModel(state, {
+        rules: state.roleModel.rules.filter(r => r.id !== payload.id),
+      });
+
+    default:
+      (type) satisfies never;
+      return state;
+  }
+};
+
+const patchRoleModel = (
+  { roleModel, originalRole, validationResult }: StandardEditorModel,
+  roleModelPatch: Partial<RoleEditorModel>
+): StandardEditorModel => {
+  const newRoleModel = { ...roleModel, ...roleModelPatch };
+  return {
+    roleModel: newRoleModel,
+    originalRole,
+    isDirty: hasModifiedFields(newRoleModel, originalRole),
+    validationResult: validateRoleEditorModel(
+      newRoleModel,
+      roleModel,
+      validationResult
+    ),
+  };
+};
+
+/**
+ * Creates a new model state given existing state and a patch to
+ * RoleEditorModel. Validates the state and recognizes whether it's dirty (i.e.
+ * changed from the original).
+ */
+const updateRoleModel = (
```

Comment:
Is there any difference between these two functions?

Reply:
Wow, good catch! Thanks :)